File: problems/024_lexicographic_permutations/lexicographic_permutations.go
// lexicographic_permutations.go
// Go reference solution using factoradic indexing. Final answer redacted.
package main

import (
    "fmt"
)

func buildFact(n int) []uint64 {
    f := make([]uint64, n+1)
    f[0] = 1
    for i := 1; i <= n; i++ {
        f[i] = f[i-1] * uint64(i)
    }
    return f
}

func nthPerm(items []rune, n1 uint64) string {
    m := len(items)
    fact := buildFact(m)
    k := n1 - 1 // zero-based
    out := make([]rune, 0, m)
    for r := m; r >= 1; r-- {
        f := fact[r-1]
        idx := int(k / f)
        k = k % f
        out = append(out, items[idx])
        items = append(items[:idx], items[idx+1:]...)
    }
    return string(out)
}

func main() {
    // sanity
    if nthPerm([]rune{'0','1','2'}, 3) != "102" {
        panic("sanity check failed")
    }
    digits := []rune{'0','1','2','3','4','5','6','7','8','9'}
    millionth := nthPerm(digits, 1_000_000)
    // fmt.Println(millionth) // TODO: reveal locally
    fmt.Println("Millionth permutation of 0..9 = [redacted]")
    _ = millionth
}

File: problems/024_lexicographic_permutations/lexicographic_permutations.cpp
// lexicographic_permutations.cpp
// C++ reference solution using factoradic indexing. Final answer redacted.
#include <bits/stdc++.h>
using namespace std;

static vector<unsigned long long> build_fact(size_t n) {
    vector<unsigned long long> f(n + 1, 1);
    for (size_t i = 1; i <= n; ++i) f[i] = f[i-1] * i;
    return f;
}

static string nth_perm(vector<char> items, unsigned long long n1) {
    size_t m = items.size();
    auto fact = build_fact(m);
    unsigned long long k = n1 - 1; // zero-based
    string out; out.reserve(m);
    for (size_t r = m; r >= 1; --r) {
        unsigned long long f = fact[r - 1];
        size_t idx = static_cast<size_t>(k / f);
        k %= f;
        out.push_back(items[idx]);
        items.erase(items.begin() + idx);
    }
    return out;
}

int main() {
    // sanity
    vector<char> small = {'0','1','2'};
    if (nth_perm(small, 3) != string("102")) return 1;

    vector<char> digits;
    for (char c='0'; c<='9'; ++c) digits.push_back(c);
    string millionth = nth_perm(digits, 1'000'000ULL);

    // cout << millionth << "\n"; // TODO: reveal locally
    cout << "Millionth permutation of 0..9 = [redacted]\n";
    (void)millionth;
    return 0;
}
